test/UniswapV3Staker.unit.spec.ts

```diff
@@ -846,6 +836,220 @@ describe('UniswapV3Staker.unit', async () => {
+            amountDesired,
+            context.nft.address
+          )
+          await context.nft.increaseLiquidity({
+            tokenId,
+            amount0Desired: amountDesired,
+            amount1Desired: amountDesired,
+            amount0Min: 0,
+            amount1Min: 0,
+            deadline: (await blockTimestamp()) + 1000,
+          })
+
+          await Time.set(timestamps.startTime + 500)
+          expectedReward = BN('20249999999999999999')
+        })
+
+        it('adjusts rewards amount for staker', async () => {
+          const rewardBefore = await context.staker.rewards(
+            context.rewardToken.address,
+            lpUser0.address
+          )
+          await subject()
+          const rewardAfter = await context.staker.rewards(
+            context.rewardToken.address,
+            lpUser0.address
+          )
+          expect(rewardAfter).to.eq(rewardBefore.add(expectedReward))
+        })
+
+        it('emits a StakeUpdated event', async () => {
+          const stakeBefore = await context.staker.stakes(tokenId, incentiveId)
+          await expect(subject())
+            .to.emit(context.staker, 'StakeUpdated')
+            .withArgs(
+              tokenId,
+              stakeBefore.liquidity.mul(2),
+              incentiveId,
+              expectedReward
+            )
+        })
+
+        it('updates the stake appropriately', async () => {
+          const stakeBefore = await context.staker.stakes(tokenId, incentiveId)
+          const secondsInitialBefore =
+            stakeBefore.secondsPerLiquidityInitialX128
+          const liquidityBefore = stakeBefore.liquidity
+
+          await subject()
+
+          const stakeAfter = await context.staker.stakes(tokenId, incentiveId)
+          const secondsInitialAfter = stakeAfter.secondsPerLiquidityInitialX128
+          const liquidityAfter = stakeAfter.liquidity
+
+          expect(secondsInitialBefore).to.be.lt(secondsInitialAfter)
+          // doubled liquidity between updates
+          expect(liquidityAfter).to.equal(liquidityBefore.mul(2))
+          expect(stakeBefore.exists).to.equal(stakeAfter.exists)
+        })
+
+        it('updates the incentive appropriately', async () => {
+          const incentivePrev = await context.staker.incentives(incentiveId)
+          const rewardsUnclaimedPrev = incentivePrev.totalRewardUnclaimed
+          const secondsClaimedPrev = incentivePrev.totalSecondsClaimedX128
+
+          await subject()
+
+          const incentiveCurrent = await context.staker.incentives(incentiveId)
+          const rewardsUnclaimedCurrent = incentiveCurrent.totalRewardUnclaimed
+          const secondsClaimedCurrent = incentiveCurrent.totalSecondsClaimedX128
+
+          expect(rewardsUnclaimedPrev.sub(rewardsUnclaimedCurrent)).to.eq(
+            expectedReward
+          )
+          expect(secondsClaimedPrev).to.equal(0)
+
+          // calculation is sometimes off by one
```

Comment:
ok this is legitimately hacky and embarrassing. `closeTo` (which is also unideal) wasn't functioning properly with BigNumber. Despite exact timestamps, sometimes figures are off by 1.

Reply:
See `beWithin.ts`